lint-rules.md

```diff
@@ -0,0 +1,1280 @@
+
+call samtools.quickcheck { input: bam }
+
+call util.compression_integrity { input:
+    bgzipped_file = bam,
+}
+
+call markdups_post_wf.markdups_post { input:
+    markdups_bam = select_first([
+        markdups.duplicate_marked_bam,
+        "undefined",
+    ]),
+    markdups_bam_index = select_first([
+        markdups.duplicate_marked_bam_index,
+        "undefined",
+    ]),
+    coverage_beds,
+    coverage_labels = parse_input.labels,
+    prefix = post_subsample_prefix + ".MarkDuplicates",
+}
+```
+
+### `snake_case`
+
+Variables, tasks, and workflows should be in lowercase ["snake_case"](https://en.wikipedia.org/wiki/Snake_case).
+
+**Group**: `naming`
+
+### `pascal_case`
+
+Declared structs should be in ["PascalCase"](https://www.theserverside.com/definition/Pascal-case).
+
+**Group**: `naming`
+
+### `indentation`
+
+Indentation should be 4 spaces and never tab literals.
+
+### `unwanted_whitespace`
+
+No whitespace on empty lines. No whitespace at the end of lines.
+
+### `one_empty_line`
+
+At most one empty line in a row. There should never be 2 or more blank lines in a row.
+
+**Group**: `spacing`
+
+### `newline_eof`
+
+The file must end with a newline.
+
+**Group**: `spacing`
+
+### `comment_whitespace`
+
+Comments on the same line as code should have 2 spaces before the `#` and one space before the comment text. Comments on their own line should match the indentation level around them and have one space between the `#` and the comment text. Keep in mind that even comments must be kept below the 90 character width limit.
+
+**Group**: `spacing`
+
+#### Example
+
+Good:
+
+```wdl
+    input {
+        Int x  # TODO better names here?
+        Int y
+    }
+
+    # Next do some math
+    Int z = x + y  # simple addition
+```
+
+Bad:
+
+```wdl
+    input {
+        Int x # Only one space before the comment. There should be two.
+        Int y  #There should be a space following the pound sign before the text.
+    }
+
+# This comment has an incorrect level of indentation
+    Int z = x + y# NO spaces between code and comment is hard to read.
+```
+
+### `double_quotes`
+
+All quotes should be double quotes.
+
+**Group**: `naming`
+
+#### Example
+
+Good:
+
+```wdl
+    output {
+        File md5sum = "path/to/file.txt"
+        String example = "Hello, I am a \"double quoted\" string!"
+    }
+
+    runtime {
+        memory: "4 GB"
+        disks: "~{disk_size_gb} GB"
+        container: "docker://ghcr.io/stjudecloud/util@sha256:c0583fe91d3e71fcfba58e2a57beb3420c7e907efd601f672fb5968086cd9acb"  # tag: 1.3.0
+        maxRetries: 1
+    }
+```
+
+Bad:
+
+```wdl
+    output {
+        File md5sum = 'path/to/file.txt'
+    }
+
+    runtime {
+        memory: '4 GB'
+        disks: '~{disk_size_gb} GB'
+        container: 'docker://ghcr.io/stjudecloud/util@sha256:c0583fe91d3e71fcfba58e2a57beb3420c7e907efd601f672fb5968086cd9acb'  # tag: 1.3.0
+        maxRetries: 1
+    }
+```
+
+### `trailing_comma`
```

Comment:
- Not sure I like this generally. I think it looks better to not have the trailing comma for lists, but understand I could just turn off the lint if I don't like it

Reply:
I think there are some list-like structures that `miniwdl check` errors on if a trailing comma is included.

Reply:
Personally, I think trailing commas look a lot cleaner and are the more visually appealing style. Trailing commas also give me less friction when actually coding. Of course those are highly subjective opinions, and anyone could disagree.

I think this is another case where there is no "best" or "objective" answer. I think the best way to settle this sort of arbitrary debate is something democratic.

Idea:
1. What if we chose some selection of WDL repos (simple list would be the repos Clay collected for `wdl-gauntlet`)
2. checked them for consistency on commas (and any other "arbitrary" lint rules we're deciding on)
3. grouped the repos by "position" on each proposed rule
4. Summed each voting blocks' number of GitHub stars
5. The rule with the most votes becomes the default for `sprocket`!

Reply:
I think some method like this makes sense, though the "samples" may not be independent, e.g. different Broad repos probably influence different Broad repos to follow similar style, etc. Maybe it would make sense to advertise whatever method you decide so other folks can chime in ideas?